core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -299,9 +301,18 @@ NodeShardsResult buildVersionBasedNodeShardsResult(ShardRouting shard, boolean m
                 continue;
             }
 
-            // no version means it does not exists, which is what the API returns, and what we expect to
             if (nodeShardState.storeException() == null) {
-                logger.trace("[{}] on node [{}] has version [{}] of shard", shard, nodeShardState.getNode(), version);
+                if (version == ShardStateMetaData.NO_VERSION && nodeShardState.allocationId() == null) {
+                    logger.trace("[{}] on node [{}] has no shard state information", shard, nodeShardState.getNode());
+                } else if (version != ShardStateMetaData.NO_VERSION) {
+                    assert nodeShardState.allocationId() == null : "Allocation id and legacy version cannot be both present";
+                    logger.trace("[{}] on node [{}] has version [{}] of shard", shard, nodeShardState.getNode(), version);
+                } else {
+                    // shard was already selected in a 3.x cluster as best candidate for recovery but did not make it to STARTED state
```

Comment:
if I understand this correctly, this part is relevant where we assigned a  primary after a cluster upgrade and the shard initialized (and wrote a new state file) but we never got around to activating it before crushing again.  if that's correct, can you add this to the comment?